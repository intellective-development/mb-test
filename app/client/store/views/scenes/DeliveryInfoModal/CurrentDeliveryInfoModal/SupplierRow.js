// @flow

import * as React from 'react';
import { connect } from 'react-redux';
import i18n from 'store/localization';
import { address_helpers } from 'store/business/address';
import { delivery_method_constants, delivery_method_helpers } from 'store/business/delivery_method';
import { supplier_actions, supplier_helpers, supplier_selectors } from 'store/business/supplier';
import type { Supplier } from 'store/business/supplier';

import { MBText, MBButton, MBIcon, MBTouchable } from '../../../elements';
import DeliveryMethodIcon from '../../../compounds/DeliveryMethodIcon';

type SupplierRowProps = {supplier: Supplier, selected_delivery_method_id: number, selectDeliveryMethod: Function, deliveryInfoRouteTo: Function};
const SupplierRow = ({supplier, selected_delivery_method_id, selectDeliveryMethod, deliveryInfoRouteTo}: SupplierRowProps) => {
  const has_pickup = supplier.delivery_methods.some(dm => dm.type === delivery_method_constants.PICKUP);

  return (
    <div className="currdel__sr__container">
      <div className="currdel__sr__header">
        <div className="currdel__sr__supplier">
          <MBText.H4 className="currdel__supplier-name">{supplier.name}</MBText.H4>
          <MBTouchable
            disabled={!has_pickup}
            onClick={() => deliveryInfoRouteTo('supplier_map', {supplier_id: supplier.id})}>
            <MBText.H5 className="currdel__supplier-location">
              {address_helpers.formatStreetAndCity(supplier.address)}
              <SupplierDistance supplier={supplier} is_hidden={!has_pickup} />
            </MBText.H5>
          </MBTouchable>
        </div>
        <ChangeSupplierButton
          is_hidden={!supplier_helpers.hasAlternatives(supplier)}
          onClick={() => deliveryInfoRouteTo('supplier_switching', {supplier_id: supplier.id})} />
      </div>
      {supplier.delivery_methods.map((delivery_method) => (
        <DeliveryMethodRow
          key={delivery_method.id}
          delivery_method={delivery_method}
          active={delivery_method.id === selected_delivery_method_id}
          selectDeliveryMethod={() => selectDeliveryMethod(supplier.id, delivery_method.id)} />
      ))}
    </div>
  );
};

const SupplierDistance = ({supplier, is_hidden}) => {
  if (is_hidden) return null;

  return (
    <MBText.Span>
      ãƒ»{supplier_helpers.formatDistance(supplier)}
    </MBText.Span>
  );
};

const ChangeSupplierButton = ({is_hidden, onClick}) => {
  if (is_hidden) return null;

  // this component renders both text and an icon, and we leave it up to the styles to display the correct one for a given screen size
  return (
    <MBButton
      type="hollow"
      size="small"
      className="currdel__sr__change-button"
      onClick={onClick}>
      <MBText.Span className="currdel__sr__change-button__text">Change</MBText.Span>
      <MBIcon name="mobile.pencil" className="currdel__sr__change-button__icon" />
    </MBButton>
  );
};

const DeliveryMethodRow = ({delivery_method}) => {
  let closed_el;
  if (delivery_method_helpers.isClosed(delivery_method)){
    closed_el = (
      <MBText.Span>
        <MBText.Span className="currdel__dm-delivery-expectation--closed">
          {i18n.t('ui.body.delivery_info.delivery_method_closed')}
        </MBText.Span>
        {' - '}
      </MBText.Span>
    );
  }

  return (
    <div className="currdel__dm-container" >
      <DeliveryMethodIcon
        delivery_method_type={delivery_method.type}
        width={30}
        height={30}
        className="currdel__dm-icon" />
      <div>
        <MBText.H4 className="currdel__dm-name">
          {delivery_method_helpers.displayName(delivery_method)}
        </MBText.H4>
        <MBText.H5 className="currdel__dm-delivery-expectation">
          {closed_el}
          {delivery_method_helpers.formatNextDelivery(delivery_method)}
        </MBText.H5>
      </div>
    </div>
  );
};

const SupplierRowSTP = (state, {supplier}) => ({
  selected_delivery_method_id: supplier_selectors.selectedDeliveryMethods(state)[supplier.id]
});
const SupplierRowDTP = {selectDeliveryMethod: supplier_actions.selectDeliveryMethod};
const SupplierRowContainer = connect(SupplierRowSTP, SupplierRowDTP)(SupplierRow);

export default SupplierRowContainer;
