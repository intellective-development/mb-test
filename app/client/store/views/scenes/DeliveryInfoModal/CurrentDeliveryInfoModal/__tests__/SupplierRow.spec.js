import * as React from 'react';
import supplier_factory from 'client/store/business/supplier/__tests__/supplier.factory';
import dm_factory from '@minibar/store-business/src/delivery_method/__tests__/delivery_method.factory';

import TestProvider from 'store/views/__tests__/utils/TestProvider';
import SupplierRow from '../SupplierRow';

describe('SupplierRow', () => {
  it('renders', () => {
    const supplier = supplier_factory.build('with_delivery_methods');

    expect(render(
      <TestProvider>
        <SupplierRow supplier={supplier} />
      </TestProvider>
    )).toMatchSnapshot();
  });

  it('renders a supplier that has a pickup delivery method', () => {
    const delivery_methods = [dm_factory.build('pickup'), dm_factory.build('shipped')];
    const supplier = supplier_factory.build({delivery_methods});

    expect(render(
      <TestProvider>
        <SupplierRow supplier={supplier} />
      </TestProvider>
    )).toMatchSnapshot();
  });
});
