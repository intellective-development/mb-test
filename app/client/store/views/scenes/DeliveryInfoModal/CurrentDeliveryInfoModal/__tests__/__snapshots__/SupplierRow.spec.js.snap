// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SupplierRow renders 1`] = `
<div
  class="currdel__sr__container"
>
  <div
    class="currdel__sr__header"
  >
    <div
      class="currdel__sr__supplier"
    >
      <h4
        class="elMBText elMBText_ResetSpacing currdel__supplier-name"
      >
        Manhattan Wine
      </h4>
      <div
        class="el-touchable el-touchable--disabled"
        role="button"
        tabindex="-1"
      >
        <h5
          class="elMBText elMBText_ResetSpacing currdel__supplier-location"
        >
           123 Jump Street, New York, New York
        </h5>
      </div>
    </div>
  </div>
  <div
    class="currdel__dm-container"
  >
    <img
      alt="on_demand"
      class="cmDeliveryMethodIcon currdel__dm-icon"
      height="30"
      src="/assets/components/compounds/delivery_method_icon/on_demand.svg"
      width="30"
    />
    <div>
      <h4
        class="elMBText elMBText_ResetSpacing currdel__dm-name"
      >
        Local Delivery
      </h4>
      <h5
        class="elMBText elMBText_ResetSpacing currdel__dm-delivery-expectation"
      >
        <span
          class="elMBText elMBText_ResetSpacing"
        >
          <span
            class="elMBText elMBText_ResetSpacing currdel__dm-delivery-expectation--closed"
          >
            Closed
          </span>
           - 
        </span>
        Next available: 1 hour or less
      </h5>
    </div>
  </div>
  <div
    class="currdel__dm-container"
  >
    <img
      alt="on_demand"
      class="cmDeliveryMethodIcon currdel__dm-icon"
      height="30"
      src="/assets/components/compounds/delivery_method_icon/on_demand.svg"
      width="30"
    />
    <div>
      <h4
        class="elMBText elMBText_ResetSpacing currdel__dm-name"
      >
        Local Delivery
      </h4>
      <h5
        class="elMBText elMBText_ResetSpacing currdel__dm-delivery-expectation"
      >
        <span
          class="elMBText elMBText_ResetSpacing"
        >
          <span
            class="elMBText elMBText_ResetSpacing currdel__dm-delivery-expectation--closed"
          >
            Closed
          </span>
           - 
        </span>
        Next available: 1 hour or less
      </h5>
    </div>
  </div>
</div>
`;

exports[`SupplierRow renders a supplier that has a pickup delivery method 1`] = `
<div
  class="currdel__sr__container"
>
  <div
    class="currdel__sr__header"
  >
    <div
      class="currdel__sr__supplier"
    >
      <h4
        class="elMBText elMBText_ResetSpacing currdel__supplier-name"
      >
        Manhattan Wine
      </h4>
      <div
        class="el-touchable"
        role="button"
        tabindex="-1"
      >
        <h5
          class="elMBText elMBText_ResetSpacing currdel__supplier-location"
        >
           123 Jump Street, New York, New York
          <span
            class="elMBText elMBText_ResetSpacing"
          >
            ãƒ»1.1 mi
          </span>
        </h5>
      </div>
    </div>
  </div>
  <div
    class="currdel__dm-container"
  >
    <img
      alt="pickup"
      class="cmDeliveryMethodIcon currdel__dm-icon"
      height="30"
      src="/assets/components/compounds/delivery_method_icon/pickup.svg"
      width="30"
    />
    <div>
      <h4
        class="elMBText elMBText_ResetSpacing currdel__dm-name"
      >
        In-Store Pickup
      </h4>
      <h5
        class="elMBText elMBText_ResetSpacing currdel__dm-delivery-expectation"
      >
        <span
          class="elMBText elMBText_ResetSpacing"
        >
          <span
            class="elMBText elMBText_ResetSpacing currdel__dm-delivery-expectation--closed"
          >
            Closed
          </span>
           - 
        </span>
        Next available: 1 hour or less
      </h5>
    </div>
  </div>
  <div
    class="currdel__dm-container"
  >
    <img
      alt="shipped"
      class="cmDeliveryMethodIcon currdel__dm-icon"
      height="30"
      src="/assets/components/compounds/delivery_method_icon/shipped.svg"
      width="30"
    />
    <div>
      <h4
        class="elMBText elMBText_ResetSpacing currdel__dm-name"
      >
        Shipping
      </h4>
      <h5
        class="elMBText elMBText_ResetSpacing currdel__dm-delivery-expectation"
      >
        <span
          class="elMBText elMBText_ResetSpacing"
        >
          <span
            class="elMBText elMBText_ResetSpacing currdel__dm-delivery-expectation--closed"
          >
            Closed
          </span>
           - 
        </span>
        Next available: 1 hour or less
      </h5>
    </div>
  </div>
</div>
`;
